### PR TITLE
Replace std::set with unordered_set in count distinct aggregation

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

File: QueryEngine/GroupByAndAggregate.cpp [hunk 1]
```diff
@@ -618,20 +621,31 @@ CountDistinctDescriptors init_count_distinct_descriptors(
...
+
+          if (group_by_hash_type == QueryDescriptionType::GroupByBaselineHash) {
+            const int64_t MAX_TOTAL_BITMAPS_BITS = 8 * 8 * 1000 * 1000 * 1000LL;  // 8GB
+            int64_t total_bitmaps_size = bitmap_sz_bits * group_by_slots_count / 2;
+
+            if (total_bitmaps_size <= 0 || total_bitmaps_size > MAX_TOTAL_BITMAPS_BITS) {
+              count_distinct_impl_type = CountDistinctImplType::Hashmap;
+            }
+          } else {
+            bitmap_sz_bits = arg_range_info.max - arg_range_info.min + 1;
```
Comment (Garra1980) [inline QueryEngine/GroupByAndAggregate.cpp:633]: No need to calculate this here?

Comment (fexolm) [reply QueryEngine/GroupByAndAggregate.cpp:633]: yep, it looks like so

Comment (fexolm) [reply QueryEngine/GroupByAndAggregate.cpp:633]:
```suggestion
```

File: QueryEngine/GroupByAndAggregate.cpp [hunk 2]
```diff
@@ -618,20 +621,31 @@ CountDistinctDescriptors init_count_distinct_descriptors(
         count_distinct_impl_type = CountDistinctImplType::Bitmap;
         if (agg_info.agg_kind == kCOUNT) {
           bitmap_sz_bits = arg_range_info.max - arg_range_info.min + 1;
-          const int64_t MAX_BITMAP_BITS{8 * 1000 * 1000 * 1000LL};
-          if (bitmap_sz_bits <= 0 || bitmap_sz_bits > MAX_BITMAP_BITS) {
-            count_distinct_impl_type = CountDistinctImplType::StdSet;
+
+          if (group_by_hash_type == QueryDescriptionType::GroupByBaselineHash) {
+            const int64_t MAX_TOTAL_BITMAPS_BITS = 8 * 8 * 1000 * 1000 * 1000LL;  // 8GB
```
Comment (ienkovich) [inline QueryEngine/GroupByAndAggregate.cpp:626]: The comment about 8GB is wrong. Also, I don't get why we have different logic for different hash tables. Why do we put a total limit for the baseline hash and a per-bitmap limit for the perfect hash case?

File: QueryEngine/serialized_result_set.thrift [hunk 3]
```diff
@@ -23,7 +23,7 @@ struct TColSlotContext {
 enum TCountDistinctImplType {
   Invalid,
   Bitmap,
-  StdSet
+  Hashmap
```
Comment (ienkovich) [inline QueryEngine/serialized_result_set.thrift:26]: Why do we call it `HashMap` if it is actually a set?

File: Shared/robin_hood.h [hunk 4]
```diff
@@ -0,0 +1,2605 @@
+//                 ______  _____                 ______                _________
```
Comment (ienkovich) [inline Shared/robin_hood.h:1]: `Shared` is for OmniSci code. This one should go to `ThirdParty`

File: QueryEngine/GroupByAndAggregate.cpp [hunk 5]
```diff
@@ -618,20 +621,31 @@ CountDistinctDescriptors init_count_distinct_descriptors(
         count_distinct_impl_type = CountDistinctImplType::Bitmap;
         if (agg_info.agg_kind == kCOUNT) {
           bitmap_sz_bits = arg_range_info.max - arg_range_info.min + 1;
-          const int64_t MAX_BITMAP_BITS{8 * 1000 * 1000 * 1000LL};
-          if (bitmap_sz_bits <= 0 || bitmap_sz_bits > MAX_BITMAP_BITS) {
-            count_distinct_impl_type = CountDistinctImplType::StdSet;
+
+          if (group_by_hash_type == QueryDescriptionType::GroupByBaselineHash) {
```
Comment (ienkovich) [inline QueryEngine/GroupByAndAggregate.cpp:625]: Isn't this condition in conflict with the one above?
`if (arg_range_info.hash_type_ == QueryDescriptionType::GroupByPerfectHash`

Comment (fexolm) [reply QueryEngine/GroupByAndAggregate.cpp:625]: nope, arg_range_info.hash_type_ == perfect_hash is only mean that this column is integer or dictionary encoding string, while group_by_hash_type  is actual query hash type

File: Shared/robin_hood.h (hunk 4 above)
Comment (fexolm) [reply Shared/robin_hood.h:1]: ok

File: QueryEngine/serialized_result_set.thrift (hunk 3 above)
Comment (fexolm) [reply QueryEngine/serialized_result_set.thrift:26]: ok

**Review by ienkovich — APPROVED** [review]

**Pull request merged**